src/vm/runners/cairo_runner.rs

```diff
@@ -760,6 +762,40 @@ impl CairoRunner {
 
         Ok(())
     }
+
+    // Checks that there are enough trace cells to fill the entire memory range.
+    pub fn check_memory_usage(&self, vm: &VirtualMachine) -> Result<(), VirtualMachineError> {
+        let instance = &self.layout;
+
+        let builtins_memory_units: usize = vm
+            .builtin_runners
+            .iter()
+            .map(|(_builtin_runner_name, builtin_runner)| {
+                builtin_runner.get_allocated_memory_units(vm)
+            })
+            .collect::<Result<Vec<usize>, MemoryError>>()?
+            .iter()
+            .sum();
+
+        let builtins_memory_units = builtins_memory_units as u32;
+
+        // Out of the memory units available per step, a fraction is used for public memory, and
+        // four are used for the instruction.
+        let total_memory_units = instance._memory_units_per_step * vm.current_step as u32;
```

Comment:
I am afraid that this could overflow in some cases. Try importing `num_traits::ToPrimitive` and use the `to_u32()` method. This will return an option that should be handled